test/net_http_test.rb

```diff
@@ -0,0 +1,389 @@
+require 'test_helper'
+require 'semian/net_http'
+require 'thin'
+
+class TestNetHTTP < MiniTest::Unit::TestCase
+  class RackServer
+    def self.call(env)
+      response_code = env['REQUEST_URI'].delete("/")
+      response_code = '200' if response_code == ""
+      [response_code, {'Content-Type' => 'text/html'}, ['Success']]
+    end
+  end
+
+  PORT = 31_050
+  TOXIC_PORT = PORT + 1
+
+  DEFAULT_SEMIAN_OPTIONS = {
+    tickets: 3,
+    success_threshold: 1,
+    error_threshold: 3,
+    error_timeout: 10,
+  }
+
+  def test_semian_identifier
+    with_server do
+      Net::HTTP.start("localhost", TOXIC_PORT) do |http|
+        assert_equal "http_localhost_#{TOXIC_PORT}", http.semian_identifier
+      end
+    end
+  end
+
+  def test_trigger_open
+    with_semian_options(DEFAULT_SEMIAN_OPTIONS) do
+      with_server do
+        open_circuit!
+        uri = URI("http://localhost:#{TOXIC_PORT}/200")
+        assert_raises Net::CircuitOpenError do
+          Net::HTTP.get(uri)
+        end
+      end
+    end
+  end
+
+  def test_trigger_close_after_open
+    with_semian_options(DEFAULT_SEMIAN_OPTIONS) do
+      with_server do
+        open_circuit!
+        close_circuit!
+      end
+    end
+  end
+
+  def test_bulkheads_tickets_are_working
+    with_semian_options(DEFAULT_SEMIAN_OPTIONS) do
+      with_server do
+        ticket_count = Net::HTTP.new("localhost", TOXIC_PORT).raw_semian_options[:tickets]
+        m = Monitor.new
+        acquired_count = 0
+        ticket_count.times do
+          threads << Thread.new do
+            http = Net::HTTP.new("localhost", "#{TOXIC_PORT}")
+            http.acquire_semian_resource(adapter: :nethttp, scope: :connection) do
+              m.synchronize do
+                acquired_count += 1
+              end
+              sleep
+            end
+          end
+        end
+
+        Thread.new do
+          sleep(1) until acquired_count == ticket_count
+          http = Net::HTTP.new("localhost", "#{TOXIC_PORT}")
+          assert_raises Net::ResourceBusyError do
+            http.get("/")
+          end
+          threads.each(&:join)
+        end
+      end
+    end
+  end
```

Comment:
This test is incredibly difficult to read. It needs to be easier to understand. I'm not even going to try :)